rfcs/0051-limit-zkapp-cmds-per-block.md

```diff
@@ -0,0 +1,161 @@
+## Summary
+
+During the ITN stress testing it was noticed that daemon's memory
+consumption tends to increase dramatically after a block containing a
+large number of zkApp commands. Before appropriate optimizations can
+be developed, we need a temporary solution to prevent nodes crashing
+due to insufficient memory. The idea is to limit the number of zkApp
+commands that can be included in any single block.
+
+## Motivation
+
+By limiting the number of zkApp commands going into blocks we avoid
+the aforementioned issue until a proper solution can be devised and
+implemented. The root cause of the issue is that proofs contained
+within these commands are stored in the scan state and tend to occupy
+a lot of space. Fixing these storage issues won't affect the
+protocol, so ideally we want a workaround that doesn't affect the
+protocol either, so that at the convenient time we can turn it off
+without making a fork.
+
+## Detailed design
+
+Since the solution should not affect the protocol, it should be
+implemented at the mempool/block producer boundary. In the mempool
+there is `transactions` function, which returns a sequence of
+transactions from the mempool in the order of decreasing transaction
+fees. The block producer module just takes a number of transactions
+from that sequence and includes them. It is relatively easy to limit
+the number of zkApp commands returned by this functions.  ZkApp
+commands above a certain limit would simply be discarded (from the
+sequence, but not from the mempool).
+
+The exact number of zkApps allowed in each block should be set
+dynamically, so that we can adjust it without redeploying nodes.
+The easiest way to control it is through a command line argument
+to the node. Thus changing the setting is as easy as restarting
+the node.
+
+The setting can be stored in the mempool configuration and
```

Comment:
I don't think this should be configurable via the CLI. We should instead bake this configuration into the binary itself, and expose a dynamic switch in the ITN authorized GraphQL for our own testing purposes. The CLI flag should never be used by neither testnet nor mainnet operators. Additionally, when we do release the fix as a soft fork fast follow release, we would be removing this CLI and making a breaking change, so best not to put it in the CLI at all unless we have a good reason to do so.

Reply:
How does the CLI flag facilitate testing better than the graphql interface? We can't control what flags people set on their nodes, so you will end up with a potentially wider set of settings on the network. If you need to run a local node with different values you can always use the graphql to configure it.

Reply:
CLI flag is better in that it makes sure that the setting is in effect ever since the node starts, while with GraphQL you've got a window between startup and calling the mutation, where a block could in theory be produced. It does not enlarge the set of settings on the network anyhow, because node operators can also use GraphQL to set any limit they like.

Reply:
Very well. I reflected that in the RFC.